### ISSUE TITLE
cannot convert from 'bool' to 'volatile std::atomic_bool

### ISSUE THREAD

**Issue by simpuc** [Cylix/cpp_redis#37]
Hello,
I am using cpp_redis with:
Microsoft Visual Studio Professional 2013
Version 12.0.40629.00 Update 5
and I am having the error `cannot convert from 'bool' to 'volatile std::atomic_bool` that I can resolve with `ATOMIC_VAR_INIT(false);`

For example, from
`std::atomic_bool callback_run(false);`
to
`std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);`
Is it a good solution?
Thank you
Simone

**Comment by Cylix** [comment]
Hi,

Thank you for your report!
I feel like it is coming from the examples right? Could you provide me the file and the line where the error is happening so that I can make the appropriate changes at the right place tonight?

Best

**Comment by simpuc** [comment]
Hi,
I have also found the error  `cannot convert from 'const char [22]' to 'cpp_redis::redis_error'` that I have resolved adding `redis_error(const std::string& _Message) : std::runtime_error(_Message){};` to redis_error.hpp.
But I am not sure the solution is correct because all the EXPECT_THROW test in reply_spec.cpp fail.

And also I think io_service.cpp needs
```
#ifdef _WIN32
#include <windows.h>
#endif 
```
at the beginning.
You'll find below all my diff.

```
diff --git a/includes/cpp_redis/redis_error.hpp b/includes/cpp_redis/redis_error.hpp
index 48d3b0d..a0f18a7 100644
--- a/includes/cpp_redis/redis_error.hpp
+++ b/includes/cpp_redis/redis_error.hpp
@@ -28,6 +28,7 @@ namespace cpp_redis {
 
 class redis_error : public std::runtime_error {
 public:
+	redis_error(const std::string& _Message) : std::runtime_error(_Message){};
   using std::runtime_error::runtime_error;
   using std::runtime_error::what;
 };
diff --git a/sources/network/io_service.cpp b/sources/network/io_service.cpp
index 5f595a5..7fdf72a 100644
--- a/sources/network/io_service.cpp
+++ b/sources/network/io_service.cpp
@@ -20,6 +20,10 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+#ifdef _WIN32
+#include <windows.h>
+#endif 
+
 #include <cpp_redis/network/io_service.hpp>
 
 #ifdef _WIN32
diff --git a/sources/network/tcp_client.cpp b/sources/network/tcp_client.cpp
index 4608acb..cb87220 100644
--- a/sources/network/tcp_client.cpp
+++ b/sources/network/tcp_client.cpp
@@ -46,7 +46,7 @@ namespace network {
 tcp_client::tcp_client(const std::shared_ptr<network::io_service>& io_service)
 : m_io_service(io_service ? io_service : io_service::get_global_instance())
 , m_sock(-1)
-, m_is_connected(false)
+, m_is_connected(ATOMIC_VAR_INIT(false))
 , m_receive_handler(nullptr)
 , m_disconnection_handler(nullptr) {
   __CPP_REDIS_LOG(debug, "cpp_redis::network::tcp_client created");
diff --git a/sources/network/windows_impl/io_service.cpp b/sources/network/windows_impl/io_service.cpp
index 97a66f9..ce215af 100644
--- a/sources/network/windows_impl/io_service.cpp
+++ b/sources/network/windows_impl/io_service.cpp
@@ -31,7 +31,7 @@ namespace windows_impl {
 
 io_service::io_service(std::size_t nb_workers)
 : network::io_service(nb_workers)
-, m_should_stop(false) {
+, m_should_stop(ATOMIC_VAR_INIT(false)) {
   //! Start winsock before any other socket calls.
   WSADATA wsaData;
   int nRet = WSAStartup(0x202, &wsaData);
diff --git a/sources/redis_client.cpp b/sources/redis_client.cpp
index 0bbc505..f8362f3 100644
--- a/sources/redis_client.cpp
+++ b/sources/redis_client.cpp
@@ -27,7 +27,7 @@ namespace cpp_redis {
 
 redis_client::redis_client(const std::shared_ptr<network::io_service>& io_service)
 : m_client(io_service)
-, m_callbacks_running(0) {
+, m_callbacks_running(ATOMIC_VAR_INIT(0)) {
   __CPP_REDIS_LOG(debug, "cpp_redis::redis_client created");
 }
```

and for tests
```
diff --git a/tests/sources/spec/redis_client_spec.cpp b/tests/sources/spec/redis_client_spec.cpp
index 59fc6f3..95756ed 100644
--- a/tests/sources/spec/redis_client_spec.cpp
+++ b/tests/sources/spec/redis_client_spec.cpp
@@ -120,7 +120,7 @@ TEST(RedisClient, SyncCommitTimeout) {
   cpp_redis::redis_client client;
 
   client.connect();
-  volatile std::atomic_bool callback_exit(false);
+  volatile std::atomic_bool callback_exit = ATOMIC_VAR_INIT(false);
   client.send({"GET", "HELLO"}, [&](cpp_redis::reply&) {
     std::this_thread::sleep_for(std::chrono::seconds(1));
     callback_exit = true;
@@ -135,7 +135,7 @@ TEST(RedisClient, SyncCommitNoTimeout) {
   cpp_redis::redis_client client;
 
   client.connect();
-  std::atomic_bool callback_exit(false);
+  std::atomic_bool callback_exit = ATOMIC_VAR_INIT(false);
   client.send({"GET", "HELLO"}, [&](cpp_redis::reply&) {
     std::this_thread::sleep_for(std::chrono::seconds(1));
     callback_exit = true;
@@ -162,7 +162,7 @@ TEST(RedisClient, SendConnectedSyncCommitConnected) {
 
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   client.send({"GET", "HELLO"}, [&](cpp_redis::reply&) {
     callback_run = true;
   });
@@ -174,7 +174,7 @@ TEST(RedisClient, SendConnectedSyncCommitConnected) {
 TEST(RedisClient, SendNotConnectedSyncCommitConnected) {
   cpp_redis::redis_client client;
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   client.send({"GET", "HELLO"}, [&](cpp_redis::reply&) {
     callback_run = true;
   });
@@ -187,7 +187,7 @@ TEST(RedisClient, SendNotConnectedSyncCommitConnected) {
 TEST(RedisClient, SendNotConnectedSyncCommitNotConnectedSyncCommitConnected) {
   cpp_redis::redis_client client;
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   client.send({"GET", "HELLO"}, [&](cpp_redis::reply&) {
     callback_run = true;
   });
@@ -258,7 +258,7 @@ TEST(RedisClient, DisconnectionHandlerWithQuit) {
   cpp_redis::redis_client client;
   std::condition_variable cv;
 
-  std::atomic_bool disconnection_handler_called(false);
+  std::atomic_bool disconnection_handler_called = ATOMIC_VAR_INIT(false);
   client.connect("127.0.0.1", 6379, [&](cpp_redis::redis_client&) {
     disconnection_handler_called = true;
     cv.notify_all();
@@ -278,7 +278,7 @@ TEST(RedisClient, DisconnectionHandlerWithoutQuit) {
   cpp_redis::redis_client client;
   std::condition_variable cv;
 
-  std::atomic_bool disconnection_handler_called(false);
+  std::atomic_bool disconnection_handler_called = ATOMIC_VAR_INIT(false);
   client.connect("127.0.0.1", 6379, [&](cpp_redis::redis_client&) {
     disconnection_handler_called = true;
     cv.notify_all();
diff --git a/tests/sources/spec/redis_subscriber_spec.cpp b/tests/sources/spec/redis_subscriber_spec.cpp
index a556801..95bfdc1 100644
--- a/tests/sources/spec/redis_subscriber_spec.cpp
+++ b/tests/sources/spec/redis_subscriber_spec.cpp
@@ -151,7 +151,7 @@ TEST(RedisSubscriber, SubConnectedCommitConnected) {
   sub.connect();
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan",
     [&](const std::string&, const std::string&) {
       callback_run = true;
@@ -178,7 +178,7 @@ TEST(RedisSubscriber, SubNotConnectedCommitConnected) {
 
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan",
     [&](const std::string&, const std::string&) {
       callback_run = true;
@@ -205,7 +205,7 @@ TEST(RedisSubscriber, SubNotConnectedCommitNotConnectedCommitConnected) {
 
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan",
     [&](const std::string&, const std::string&) {
       callback_run = true;
@@ -232,7 +232,7 @@ TEST(RedisSubscriber, SubscribeSomethingPublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan",
     [&](const std::string& channel, const std::string& message) {
       EXPECT_TRUE(channel == "/chan");
@@ -262,7 +262,7 @@ TEST(RedisSubscriber, SubscribeMultiplePublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_int number_times_called(0);
+  std::atomic_int number_times_called = ATOMIC_VAR_INIT(0);
   sub.subscribe("/chan",
     [&](const std::string& channel, const std::string& message) {
       EXPECT_TRUE(channel == "/chan");
@@ -295,7 +295,7 @@ TEST(RedisSubscriber, SubscribeNothingPublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan",
     [&](const std::string&, const std::string&) {
       callback_run = true;
@@ -328,8 +328,8 @@ TEST(RedisSubscriber, MultipleSubscribeSomethingPublished) {
     }
   };
 
-  std::atomic_bool callback_1_run(false);
-  std::atomic_bool callback_2_run(false);
+  std::atomic_bool callback_1_run = ATOMIC_VAR_INIT(false);
+  std::atomic_bool callback_2_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan_1",
     [&](const std::string& channel, const std::string& message) {
       EXPECT_TRUE(channel == "/chan_1");
@@ -369,7 +369,7 @@ TEST(RedisSubscriber, PSubscribeSomethingPublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.psubscribe("/chan/*",
     [&](const std::string& channel, const std::string& message) {
       EXPECT_TRUE(channel == "/chan/hello");
@@ -399,7 +399,7 @@ TEST(RedisSubscriber, PSubscribeMultiplePublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_int number_times_called(0);
+  std::atomic_int number_times_called = ATOMIC_VAR_INIT(0);
   sub.psubscribe("/chan/*",
     [&](const std::string& channel, const std::string& message) {
       ++number_times_called;
@@ -439,7 +439,7 @@ TEST(RedisSubscriber, PSubscribeNothingPublished) {
   sub.connect();
   client.connect();
 
-  std::atomic_bool callback_run(false);
+  std::atomic_bool callback_run = ATOMIC_VAR_INIT(false);
   sub.psubscribe("/chan/*",
     [&](const std::string&, const std::string&) {
       callback_run = true;
@@ -472,8 +472,8 @@ TEST(RedisSubscriber, MultiplePSubscribeSomethingPublished) {
     }
   };
 
-  std::atomic_bool callback_1_run(false);
-  std::atomic_bool callback_2_run(false);
+  std::atomic_bool callback_1_run = ATOMIC_VAR_INIT(false);
+  std::atomic_bool callback_2_run = ATOMIC_VAR_INIT(false);
   sub.psubscribe("/chan/*",
     [&](const std::string& channel, const std::string& message) {
       EXPECT_TRUE(channel == "/chan/1");
@@ -521,8 +521,8 @@ TEST(RedisSubscriber, Unsubscribe) {
     }
   };
 
-  std::atomic_bool callback_1_run(false);
-  std::atomic_bool callback_2_run(false);
+  std::atomic_bool callback_1_run = ATOMIC_VAR_INIT(false);
+  std::atomic_bool callback_2_run = ATOMIC_VAR_INIT(false);
   sub.subscribe("/chan_1",
     [&](const std::string&, const std::string&) {
       callback_1_run = true;
@@ -562,8 +562,8 @@ TEST(RedisSubscriber, PUnsubscribe) {
     }
   };
 
-  std::atomic_bool callback_1_run(false);
-  std::atomic_bool callback_2_run(false);
+  std::atomic_bool callback_1_run = ATOMIC_VAR_INIT(false);
+  std::atomic_bool callback_2_run = ATOMIC_VAR_INIT(false);
   sub.psubscribe("/chan_1/*",
     [&](const std::string&, const std::string&) {
       callback_1_run = true;
```

**Comment by Cylix** [comment]
Ok, thanks a lot for the diffs!
Apparently coming from tests and examples.
I don't have these compilation errors on my version of VC++ (2015).
Will fix that tonight and will try to see if I can use an earlier version of VC++ from VC++ 2015.

**Comment by navossoc** [comment]
I can also reproduce that error on VS 2013 Update 5.

**Comment by Cylix** [comment]
Fixed in #42.
